docs/design/004-user-stack.md

```diff
@@ -0,0 +1,169 @@
+to work together within the [OpenTelemetry Profiling
+SIG](https://github.com/open-telemetry/community). The main git repository for
+this effort is located at
+[opentelemetry-ebpf-profiler](https://github.com/open-telemetry/opentelemetry-ebpf-profiler).
+
+You can read more about opentelemetry-ebpf-profiler:
+- https://opentelemetry.io/blog/2024/profiling/
+- https://opentelemetry.io/blog/2024/elastic-contributes-continuous-profiling-agent/
+- https://www.polarsignals.com/blog/posts/2024/08/13/parca-opentelemetry-bright-future
+
+### Enriching events with user stack trace
+
+A lot of gadgets in Inspektor Gadget reports when an application is performing
+a specific system call (e.g. the `trace_open` gadget monitors the `open` and
+`openat` system calls). This tells **what** an application is doing but not
+**why**. To answer the **why** question, it would be useful to show the user
+stack trace at the time of the system call.
+
+In this usecase, we're not running a BPF program of type `perf_event` at a
+specific frequency but instead we're running a BPF program of type kprobe or
+uprobe. But the challenges remain the same and ideally, we would build upon the
+work from OpenTelemetry.
+
+## Possible implementations
+
+### Use `BPF_MAP_TYPE_STACK_TRACE` and reimplement stack unwinding
+
+Given the difficulties, Inspektor Gadget would only implement basic support for
+programs that include frame pointers.
+
+Pros:
+- Work for simple use cases without external dependencies.
+
+Cons:
+- In practice, it would not work for most applications without frame pointers.
+  It would not work for interpreted languages.
+
+### Integrating opentelemetry-ebpf-profiler as a library
+
+Pros:
+- Like Inspektor Gadget, opentelemetry-ebpf-profiler is written in Golang for the
+  userspace parts and C for the BPF parts. So it would feel natural to import the
+  Go packages in Inspektor Gadget and reuse the BPF code in some way.
+
+Cons:
+- opentelemetry-ebpf-profiler is a big project. GitHub pulse shows that inspektor-gadget and
+  opentelemetry-ebpf-profiler have a similar amount of authors, commits, line of
+  code and stars in the last month.
+  - It would significantly increase the size of the ig executable.
+  - If several software using this at the same time, the BPF programs would be
+    loaded twice, resulting in twice more CPU time being consumed.
+- opentelemetry-ebpf-profiler is not designed as a library.
+
+Issue discussing this implementation:
+https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/33
+
+### gRPC communication between Inspektor Gadget and opentelemetry-ebpf-profiler
+
+When a gadget's BPF code wants to get a user stack trace, it would call a BPF
+program prepared by opentelemetry-ebpf-profiler returning a stack trace id.
+Then, Inspektor Gadget can request in userspace the details of that stack trace
+id.
+
+![IG / OTel profiler](004-ig-otel-profiler.png)
+
+#### Why gRPC on unix sockets?
```

Comment:
Triggering of stack unwinding requests from Inspector Gadget to OTel eBPF profiler will take some time and will be interrupted e.g. by the Linux scheduler.
OTel eBPF profiler is not capable of unwinding stack from past events. So assigning the user stack component for a `open_at` syscall of a process might not be possible.